### PR TITLE
Improved support for binary I/O

### PR DESCRIPTION
**Pull request by gabbard**

### CODE REVIEW

File: vistautils/key_value.py [hunk 1]
```diff
@@ -733,16 +807,28 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+def _doc_id_binary_source_from_params(params: Parameters) -> KeyValueSource[str, bytes]:
+    return KeyValueSource.binary_from_doc_id_to_file_map(params.existing_file("path"))
+
+
 def char_key_value_linear_source_from_params(
-    param_name: str, params: Parameters, *, eval_context: Optional[Dict] = None
+    params: Parameters,
+    *,
+    input_namespace: str = "input",
+    eval_context: Optional[Dict] = None,
```
Comment (berquist) [inline vistautils/key_value.py]:
```suggestion
    eval_context: Optional[Mapping[Any, Any]] = None,
```
for all the functions that take an `eval_context`.

File: vistautils/key_value.py [hunk 2]
```diff
@@ -565,6 +626,19 @@ def __init__(
...
 
+    @staticmethod
+    def from_parameters(params: Parameters) -> KeyValueSource[str, bytes]:
+        """
+        Construct a zip file key-value source from parameters.
+
+        The "path" parameter should be the zip file to be opened.
+
+        Currently we assume the zipfile contains a file with the IDs, which it will if it
+        were created by the default `ByteSink.zip_bytes_sink()`. Support for custom key
```
Comment (berquist) [inline vistautils/key_value.py]:
```suggestion
        were created by the default `KeyValueSink.zip_bytes_sink()`. Support for custom key
```

File: vistautils/key_value.py [hunk 3]
```diff
@@ -772,16 +858,63 @@ def char_key_value_linear_source_from_params(
...
+    called with these parameters and should return a `KeyValueLinearSource[str, bytes]`. If a
+    callable, it will be called with these parameters (and should also return a
+    KeyValueLinearSource[str, bytes]).
+
+    The type 'zip' is a shortcut for a key-value zip file.  'file-map' is a shortcut for a
+    docID to file map.
+
+    If additional imports are needed to resolve 'type', they can be specified as a Python
+    list in
+    the `imports` field.
```
Comment (berquist) [inline vistautils/key_value.py]: I can't find this in `Parameters.object_from_parameters`.

File: vistautils/key_value.py (hunk 1 above)
Comment (berquist) [reply vistautils/key_value.py]: I also think that as a rule, we should disallow the implicit `Any` in annotations.

Comment (rgabbard) [reply vistautils/key_value.py]: Added https://github.com/isi-vista/vistautils/issues/86

File: vistautils/key_value.py (hunk 3 above)
Comment (rgabbard) [reply vistautils/key_value.py]: @berquist : It was present but not documented (and also wrongly pluralized). I have pushed updated documentation in 8e792dd and 2b072be

Comment (berquist) [reply vistautils/key_value.py]: I see it now in https://github.com/isi-vista/vistautils/blob/7eb87c06d3dc31d8c29f43d932ef6b834500c3f2/vistautils/parameters.py#L536.

File: tests/test_io_utils.py [hunk 4]
```diff
@@ -148,35 +148,9 @@ def test_read_write_doc_id_to_file_map(self):
...
-            pass
-
-        argless_params = Parameters.from_mapping({"test": "ArglessTestObj"})
-        self.assertEqual(
-            ArglessTestObj(),
-            argless_params.object_from_parameters(
-                "test", ArglessTestObj, context=locals()
-            ),
-        )
+    def test_to_file_byte(self):
```
Comment (gabbard) [inline tests/test_io_utils.py]: Prefer to use pytest's automatic temp directory support, like this (outside of any class):

```python
def test_to_file_bytes(temp_dir):
   file_path = tmp_dir / "test.txt"
   ...
```

File: tests/test_io_utils.py [hunk 5]
```diff
@@ -148,35 +148,9 @@ def test_read_write_doc_id_to_file_map(self):
...
+        tmp_dir = Path(tempfile.mkdtemp())
+        file_path = tmp_dir / "test.txt"
+        byte_sink = ByteSink.to_file(file_path).to_buffer()
```
Comment (gabbard) [inline tests/test_io_utils.py]: I am not sure exactly what is happening here - `to_buffer` should be a static method on the class `ByteSink` which produces a different type of `ByteSink`. Instead you want something like 

```
byte_sink = ByteSink.to_file(file_path)
byte_sink.write("hello\n\nworld".encode("utf-8"))
self.assertEqual("hello\n\nworld", ByteSource.from_file(file_path).read().decode("utf-8"))
```

**Pull request merged**